src/DIRAC/Core/Utilities/Subprocess.py

```diff
@@ -241,60 +241,58 @@ def __selectFD(self, readSeq, timeout=False):
         events = sel.select(timeout=timeout or self.timeout or None)
         return [key.fileobj for key, event in events if event & selectors.EVENT_READ]
 
-    def __killPid(self, pid, sig=9):
-        """send signal :sig: to process :pid:
-
-        :param int pid: process id
-        :param int sig: signal to send, default 9 (SIGKILL)
-        """
+    def __killProcess(self, process):
+        """Kills a Process with SIGTERM and returns a (gone, alive) tuple"""
+        self.log.verbose(f"Sending kill signal to PID {process.pid}")
         try:
-            os.kill(pid, sig)
-        except Exception as x:
-            if str(x) != "[Errno 3] No such process":
-                self.log.exception("Exception while killing timed out process")
-                raise x
+            process.send_signal(signal.SIGTERM)
+        except psutil.NoSuchProcess:
+            return ([], [])
+        return psutil.wait_procs([process])
```

Comment:
Isn't there a risk of waiting forever here?
Especially knowing that we had issue to finish the processes before:

> From what I understand, Subprocess.killChild() logic fails to close the stdout/stderr pipes, so the ExecutionThread never returns and the JobWrapper spins forever.`

https://github.com/DIRACGrid/DIRAC/pull/8179

According to the documentation, setting a timeout could prevent that: 

> This function will return as soon as all processes terminate or when timeout (seconds) occurs. Differently from [Process.wait()](https://psutil.readthedocs.io/en/latest/index.html#psutil.Process.wait) it will not raise [TimeoutExpired](https://psutil.readthedocs.io/en/latest/index.html#psutil.TimeoutExpired) if timeout occurs.`

https://psutil.readthedocs.io/en/latest/index.html#psutil.wait_procs

Could you try to add a `timeout` and see if it changes the behavior of the `JobWrapper` test for instance?

Reply:
If you introduce the timeout, it means that you can potentially have some `alive` threads, and here you would need to kill them for good.


```suggestion        
        terminated_gone, terminated_alive = psutil.wait_procs([process], timeout=timeout)
    
        if terminated_alive:
            self.log.info(f"Process {process.pid} did not terminate after SIGTERM, escalating to SIGKILL")
            for p in terminated_alive:
                try:
                    p.kill()
                except psutil.NoSuchProcess:
                    pass
            killed_gone, killed_alive = psutil.wait_procs(terminated_alive, timeout=5)
            gone.extend(killed_gone)
            return (gone, killed_alive)
    
    return (gone, alive)
```

You could potentially print additional details if a terminated + killed process is still alive after the timeout.

Reply:
But then it looks like we are kind of back to what we had here actually: https://github.com/aldbr/DIRAC/blob/940f06c8dad164a50e12988a94ddd49e67d51aa2/src/DIRAC/Core/Utilities/Subprocess.py#L266-L295

May be we should print additional details to understand why certain processes are not killed as they should?

Reply:
(Actually the issue was likely caused by the first `if` condition, which is supposed to be `True` if we use `start_new_session=True`)

Reply:
I just did these changes but the JobWrapper test did not change.